// prettier-plugin-jsdoc/lib/tags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.YIELDS = exports.VERSION = exports.TYPEDEF = exports.TYPE = exports.TODO = exports.THROWS = exports.TEMPLATE = exports.SINCE = exports.SEE = exports.RETURNS = exports.PROVIDES_MODULE = exports.PROPERTY = exports.PRIVATE = exports.PARAM = exports.OVERRIDE = exports.MEMBEROF = exports.MEMBER = exports.FUNCTION = exports.FLOW = exports.FIRES = exports.FILE = exports.EXTERNAL = exports.EXTENDS = exports.EXAMPLE = exports.DESCRIPTION = exports.DEPRECATED = exports.DEFAULT = exports.CONSTANT = exports.CLASS = exports.CATEGORY = exports.CALLBACK = exports.AUTHOR = exports.AUGMENTS = exports.ASYNC = exports.ABSTRACT = void 0;
const ABSTRACT = "abstract";
exports.ABSTRACT = ABSTRACT;
const ASYNC = "async";
exports.ASYNC = ASYNC;
const AUGMENTS = "augments";
exports.AUGMENTS = AUGMENTS;
const AUTHOR = "author";
exports.AUTHOR = AUTHOR;
const CALLBACK = "callback";
exports.CALLBACK = CALLBACK;
const CATEGORY = "category";
exports.CATEGORY = CATEGORY;
const CLASS = "class";
exports.CLASS = CLASS;
const CONSTANT = "constant";
exports.CONSTANT = CONSTANT;
const DEFAULT = "default";
exports.DEFAULT = DEFAULT;
const DEPRECATED = "deprecated";
exports.DEPRECATED = DEPRECATED;
const DESCRIPTION = "description";
exports.DESCRIPTION = DESCRIPTION;
const EXAMPLE = "example";
exports.EXAMPLE = EXAMPLE;
const EXTENDS = "extends";
exports.EXTENDS = EXTENDS;
const EXTERNAL = "external";
exports.EXTERNAL = EXTERNAL;
const FILE = "file";
exports.FILE = FILE;
const FIRES = "fires";
exports.FIRES = FIRES;
const FLOW = "flow";
exports.FLOW = FLOW;
const FUNCTION = "function";
exports.FUNCTION = FUNCTION;
const MEMBER = "member";
exports.MEMBER = MEMBER;
const MEMBEROF = "memberof";
exports.MEMBEROF = MEMBEROF;
const OVERRIDE = "override";
exports.OVERRIDE = OVERRIDE;
const PARAM = "param";
exports.PARAM = PARAM;
const PRIVATE = "private";
exports.PRIVATE = PRIVATE;
const PROPERTY = "property";
exports.PROPERTY = PROPERTY;
const PROVIDES_MODULE = "providesModule";
exports.PROVIDES_MODULE = PROVIDES_MODULE;
const RETURNS = "returns";
exports.RETURNS = RETURNS;
const SEE = "see";
exports.SEE = SEE;
const SINCE = "since";
exports.SINCE = SINCE;
const TEMPLATE = "template";
exports.TEMPLATE = TEMPLATE;
const THROWS = "throws";
exports.THROWS = THROWS;
const TODO = "todo";
exports.TODO = TODO;
const TYPE = "type";
exports.TYPE = TYPE;
const TYPEDEF = "typedef";
exports.TYPEDEF = TYPEDEF;
const VERSION = "version";
exports.VERSION = VERSION;
const YIELDS = "yields";
exports.YIELDS = YIELDS;
