// serialize-query-params/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var withDefault_1 = require("./withDefault");
exports.withDefault = withDefault_1.withDefault;
var serialize_1 = require("./serialize");
exports.encodeDate = serialize_1.encodeDate;
exports.decodeDate = serialize_1.decodeDate;
exports.encodeBoolean = serialize_1.encodeBoolean;
exports.decodeBoolean = serialize_1.decodeBoolean;
exports.encodeNumber = serialize_1.encodeNumber;
exports.decodeNumber = serialize_1.decodeNumber;
exports.encodeString = serialize_1.encodeString;
exports.decodeString = serialize_1.decodeString;
exports.decodeEnum = serialize_1.decodeEnum;
exports.encodeJson = serialize_1.encodeJson;
exports.decodeJson = serialize_1.decodeJson;
exports.encodeArray = serialize_1.encodeArray;
exports.decodeArray = serialize_1.decodeArray;
exports.encodeNumericArray = serialize_1.encodeNumericArray;
exports.decodeNumericArray = serialize_1.decodeNumericArray;
exports.encodeDelimitedArray = serialize_1.encodeDelimitedArray;
exports.decodeDelimitedArray = serialize_1.decodeDelimitedArray;
exports.encodeDelimitedNumericArray = serialize_1.encodeDelimitedNumericArray;
exports.decodeDelimitedNumericArray = serialize_1.decodeDelimitedNumericArray;
exports.encodeObject = serialize_1.encodeObject;
exports.decodeObject = serialize_1.decodeObject;
exports.encodeNumericObject = serialize_1.encodeNumericObject;
exports.decodeNumericObject = serialize_1.decodeNumericObject;
var params_1 = require("./params");
exports.StringParam = params_1.StringParam;
exports.NumberParam = params_1.NumberParam;
exports.ObjectParam = params_1.ObjectParam;
exports.ArrayParam = params_1.ArrayParam;
exports.NumericArrayParam = params_1.NumericArrayParam;
exports.JsonParam = params_1.JsonParam;
exports.DateParam = params_1.DateParam;
exports.DateTimeParam = params_1.DateTimeParam;
exports.BooleanParam = params_1.BooleanParam;
exports.NumericObjectParam = params_1.NumericObjectParam;
exports.DelimitedArrayParam = params_1.DelimitedArrayParam;
exports.DelimitedNumericArrayParam = params_1.DelimitedNumericArrayParam;
exports.createEnumParam = params_1.createEnumParam;
var updateLocation_1 = require("./updateLocation");
exports.updateLocation = updateLocation_1.updateLocation;
exports.updateInLocation = updateLocation_1.updateInLocation;
exports.transformSearchStringJsonSafe = updateLocation_1.transformSearchStringJsonSafe;
var encodeQueryParams_1 = require("./encodeQueryParams");
exports.encodeQueryParams = encodeQueryParams_1.encodeQueryParams;
var decodeQueryParams_1 = require("./decodeQueryParams");
exports.decodeQueryParams = decodeQueryParams_1.decodeQueryParams;
if (process.env.NODE_ENV !== 'production' && typeof require === 'function') {
    /*
     * run checks to ensure a valid version of query-string is installed
     * see https://github.com/pbeshai/use-query-params/issues/127 for discussion
     */
    var queryStringVersion = require('query-string/package.json').version;
    // simple check of versions since we don't anticipate any new minor v5s and
    // don't want to require the semver package as a dependency for just a simple
    // dev check.
    var validQueryStringInstalled = /^5.1.[1-9][0-9]*/.test(queryStringVersion) ||
        /^6\./.test(queryStringVersion);
    if (!validQueryStringInstalled) {
        throw new Error("serialize-query-params requires query-string ^5.1.1 || ^6, " +
            ("but " + queryStringVersion + " is installed. Note: you may also ") +
            "see this message if you're using use-query-params.");
    }
}
